app/src/main/java/org/p2p/wallet/striga/offramp/ui/StrigaOffRampPresenter.kt

```diff
@@ -0,0 +1,344 @@
+import org.p2p.wallet.striga.offramp.mappers.StrigaOffRampMapper
+import org.p2p.wallet.striga.offramp.mappers.StrigaOffRampSwapWidgetMapper
+import org.p2p.wallet.striga.offramp.models.StrigaOffRampButtonState
+import org.p2p.wallet.striga.offramp.models.StrigaOffRampRateState
+import org.p2p.wallet.striga.offramp.models.StrigaOffRampTokenState
+import org.p2p.wallet.striga.offramp.models.StrigaOffRampTokenType
+import org.p2p.wallet.tokenservice.TokenServiceCoordinator
+import org.p2p.wallet.tokenservice.UserTokensState
+
+private const val TAG = "StrigaOffRampPresenter"
+private val DEFAULT_INITIAL_AMOUNT = BigDecimal.ZERO
+
+class StrigaOffRampPresenter(
+    private val connectionManager: ConnectionManager,
+    private val interactor: StrigaOffRampInteractor,
+    private val tokenServiceCoordinator: TokenServiceCoordinator,
+    private val strigaOffRampMapper: StrigaOffRampMapper,
+    private val swapWidgetMapper: StrigaOffRampSwapWidgetMapper,
+    private val rateTickerMapper: SwapRateTickerMapper,
+) : BasePresenter<View>(), Presenter {
+
+    private val timber: Timber.Tree = Timber.tag(TAG)
+
+    private val viewState = MutableStateFlow(
+        StrigaOffRampViewState(
+            exchangeRateState = SwapRateTickerState.Loading,
+            buttonState = StrigaOffRampButtonState.LoadingRates,
+        )
+    )
+
+    private var rate: StrigaExchangeRate? = null
+    private var balance: BigDecimal = DEFAULT_INITIAL_AMOUNT
+    private var inputAmountA: BigDecimal = DEFAULT_INITIAL_AMOUNT
+    private var inputAmountB: BigDecimal = DEFAULT_INITIAL_AMOUNT
+    private var isErrorHappened: Boolean = false
+
+    private val cannotClickAllAmount: Boolean
+        get() = isErrorHappened
+
+    override fun attach(view: View) {
+        super.attach(view)
+        interactor.startPolling(this)
+        observeUsdc()
+        observeRateChanges()
+        observeInternetState()
+        observeViewState()
+    }
+
+    override fun detach() {
+        super.detach()
+        interactor.stopPolling()
+    }
+
+    private fun observeUsdc() {
+        launch {
+            tokenServiceCoordinator.observeUserTokens()
+                .collect { handleTokenState(it) }
+        }
+    }
+
+    private fun observeRateChanges() {
+        launch {
+            interactor.observeExchangeRateState()
+                .collect { state -> handleRateChange(state) }
+        }
+    }
+
+    private fun observeInternetState() {
+        launch {
+            connectionManager.connectionStatus.collect { hasConnection ->
+                if (hasConnection) {
+                    interactor.startPolling(this)
+
+                    // restore after error
+                    if (isErrorHappened) {
+                        isErrorHappened = false
+                        tokenServiceCoordinator.refresh()
+                    }
+                } else {
+                    isErrorHappened = true
```

Comment:
I'm certain there's a more elegant solution to this, but at the moment I'm unsure of what that might be